File: domain/distiller.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Protocol

from pydantic import BaseModel, Field


class DistilledFact(BaseModel):
    subject: str
    predicate: str
    object: str
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: str
    volatility: str = "medium"


class DistilledEpisode(BaseModel):
    summary: str
    participants: list[str] = []
    entities: list[str] = []
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: str


class DistilledNote(BaseModel):
    text: str
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: str


class DistillationResult(BaseModel):
    facts: list[DistilledFact] = []
    episodes: list[DistilledEpisode] = []
    notes: list[DistilledNote] = []
    rejected: list[str] = []


class IMemoryDistiller(ABC):
    @abstractmethod
    def distill(self, text: str) -> DistillationResult:
        raise NotImplementedError

File: templates/promt/distiller.js
Extract durable memory candidates from the user message.

Return only valid JSON with this shape:
{
  "facts": [
    {
      "subject": "...",
      "predicate": "...",
      "object": "...",
      "confidence": 0.0,
      "evidence": "...",
      "volatility": "low|medium|high"
    }
  ],
  "episodes": [
    {
      "summary": "...",
      "participants": [],
      "entities": [],
      "confidence": 0.0,
      "evidence": "..."
    }
  ],
  "notes": [
    {
      "text": "...",
      "confidence": 0.0,
      "evidence": "..."
    }
  ],
  "rejected": []
}

Rules:
- Extract only information explicitly stated in the message.
- Do not infer hidden facts.
- Do not store temporary or trivial details.
- If the text contains uncertainty, lower confidence.
- Use normalized predicates like: location, preference, works_with, owns, status, skill, goal.
- Evidence must be a short quote from the input.